### PR TITLE
[Zvelo] Create a connector to download Zvelo intel

### PR DESCRIPTION
**Pull request by romain-filigran**
### Proposed changes

* First version of a connector for Zvelo Threat Intelligence Feeds

### Related issues

* https://github.com/OpenCTI-Platform/connectors/issues/2591

### Checklist

- [X] I consider the submitted work as finished
- [X] I tested the code for its functionality using different use cases
- [X] I added/update the relevant documentation (either on github or on notion)
- [X] Where necessary I refactored code to improve the overall quality

### CODE REVIEW

File: external-import/zvelo/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      # - CONNECTOR_RUN_AND_TERMINATE=False # Default False, True run connector once
+      # - CONNECTOR_SEND_TO_QUEUE=True # Default True
+      # - CONNECTOR_SEND_TO_DIRECTORY=False # Default False 
+      # - CONNECTOR_SEND_TO_DIRECTORY_PATH=CHANGEME # if CONNECTOR_SEND_TO_DIRECTORY is True, you must specify a path
+      # - CONNECTOR_SEND_TO_DIRECTORY_RETENTION=7 # Default 7, in days
+
+      # Connector's custom execution parameters
+      - ZVELO_CLIENT_ID=CHANGEME
+      - ZVELO_CLIENT_SECRET=CHANGEME
+      - ZVELO_COLLECTIONS='phish,threat,malicious'
```
Comment (Powlinett) [inline external-import/zvelo/docker-compose.yml]: ⚠️ single quotes are interpreted as part of string and produce errors like:
```
connector-zvelo-1  | {"timestamp": "2024-12-17T17:53:48.038981Z", "level": "ERROR", "name": "Zvelo", "message": "Unsupported configured: 'phish", "exc_info": "NoneType: None", "taskName": null}
connector-zvelo-1  | {"timestamp": "2024-12-17T17:53:48.039193Z", "level": "ERROR", "name": "Zvelo", "message": "Unsupported configured: malicious'", "exc_info": "NoneType: None", "taskName": null}
```

File: external-import/zvelo/src/zvelo_connector/connector.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    """
+    Specifications of the external import connector
+
+    This class encapsulates the main actions, expected to be run by any external import connector.
+    Note that the attributes defined below will be complemented per each connector type.
+    This type of connector aim to fetch external data to create STIX bundle and send it in a RabbitMQ queue.
+    The STIX bundle in the queue will be processed by the workers.
+    This type of connector uses the basic methods of the helper.
+
+    ---
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/connector.py]: You can safely remove this kind of comments, it's useful in templates only

File: external-import/zvelo/src/zvelo_connector/connector.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+    def _collect_intelligence(self, from_date) -> list:
+        """
+        Collect intelligence from the source and convert into STIX object
+        :return: List of STIX objects
+        """
+
+        # check collections configured
+        collections_str = self.config.zvelo_collections
+        if "," in collections_str:
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/connector.py]: Several comments here:
- you don't need to check if string contains ",", `collections_str.split(",")` will always return a list, even with empty string (https://docs.python.org/3/library/stdtypes.html#str.split)
- it's really great to add extra validation of supported collections
- note that config parsing and validation should belong to `ConfigConnector()` class for better separation of concerns

File: external-import/zvelo/src/zvelo_connector/connector.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        :return: None
+        """
+        self.helper.connector_logger.info(
+            "[CONNECTOR] Starting connector...",
+            {"connector_name": self.helper.connect_name},
+        )
+
+        try:
+            # Get the current state
+            now_utc = datetime.utcnow()
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/connector.py]: `datetime.utcnow()` is deprecated, please prefer to use `datetime.now(timezone.utc)` (https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow)

File: external-import/zvelo/src/zvelo_connector/connector.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+            if stix_objects is not None and len(stix_objects) != 0:
+                stix_objects_bundle = self.helper.stix2_create_bundle(stix_objects)
+                bundles_sent = self.helper.send_stix2_bundle(
+                    stix_objects_bundle, work_id=work_id
+                )
+
+                self.helper.connector_logger.info(
+                    "[CONNECTOR] Sending STIX objects to OpenCTI...",
+                    {"bundles_sent": {str(len(bundles_sent))}},
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/connector.py]: small typo: extra curly braces around `str(len(bundles_sent))` should be removed for better logs display

File: external-import/zvelo/src/zvelo_connector/connector.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+                )
+
+            # Store the current timestamp as a last run of the connector
+            self.helper.connector_logger.debug(
+                "Getting current state and update it with last run of the connector",
+                {"current_timestamp": current_timestamp},
+            )
+            current_state = self.helper.get_state()
+            current_state_datetime = now_utc.strftime("%Y-%m-%dT%H:%M:%S")
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/connector.py:143]: When I don't need a specific date string format, I personally tend to use `now_utc.isoformat()` which is easier to read and less error-prone IMO (https://docs.python.org/3/library/datetime.html#datetime.datetime.isoformat)

File: external-import/zvelo/src/zvelo_connector/connector.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+            last_run_datetime = datetime.utcfromtimestamp(current_timestamp).strftime(
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/connector.py]: I don't quite understand the purpose of `last_run_datetime`.
It's used only for logs but then the logs don't display the right `last_run` date which is actually equal to `current_state_datetime` 🙃
`last_run_datetime` should probably be removed, right?

File: external-import/zvelo/src/zvelo_connector/config_variables.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        self.duration_period = get_config_variable(
+            "CONNECTOR_DURATION_PERIOD",
+            ["connector", "duration_period"],
+            self.load,
+        )
+
+        self.zvelo_client_id = get_config_variable(
+            "ZVELO_CLIENT_ID",
+            ["zvelo", "client_id"],
+            self.load,
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/config_variables.py]: If a config variable is required, it'd be better to reflect it in the code by adding keyword arg `required=True`. That way it will fail fast and return a more meaningful error message to users.

The same applies for `zvelo_client_secret` and `zvelo_collections`.
For `zvelo_collections` you can also provide `default="phish,malicious,threat"` to stick with the README (and perform parsing/validation here as mentioned earlier).

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+            if self._is_ipv4(ip_address.get("ip")):
+                stix_related_ip_observable = stix2.IPv4Address(
+                    value=ip_address.get("ip"),
+                    object_marking_refs=[self.tlp],
+                    custom_properties={
+                        "x_opencti_created_by_ref": self.author["id"],
+                        "labels": labels,
+                    },
+                )
+            else:
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: Could it be replaced by `elif self._is_ipv6(ip_address.get("ip"))` for extra-security and explicit intention? 🤷‍♀️

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+                stix_related_ip_observable = stix2.IPv6Address(
+                    value=ip_address.get("ip"),
+                    object_marking_refs=[self.tlp],
+                    custom_properties={
+                        "x_opencti_created_by_ref": self.author["id"],
+                        "labels": labels,
+                    },
+                )
+
+            stix_relationship = stix2.Relationship(
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: Refacto opportunity here:
```
stix_relationship = self._create_relation(
      observable_id, stix_related_ip_observable.id, "related-to"
  )
```

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+            relationship_type=relation,
+            source_ref=source_id,
+            target_ref=target_id,
+            object_marking_refs=[self.tlp],
+        )
+        return stix_relationship
+
+    def convert_threat_to_stix(self, data):
+        bundle_objects = []
+        if data.get("ioc_type", None) == "url":
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: `None` can be safely removed from all occurrences of `data.get(_key_, None)` as it will return `None` anyway. It'll improve readability IMO (https://docs.python.org/3/library/stdtypes.html#dict.get)

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+                },
+            )
+            # create STIX indicator
+            pattern_value = "[domain-name:value = '" + data.get("ioc") + "']"
+            observable_type = "Domain-Name"
+
+        elif data.get("ioc_type", None) == "ip":
+            if ":" in data.get("ioc"):
+                ip_value = data.get("ioc").split(":")[0]
+                description = f"Traffic seen on port {data.get("ioc").split(":")[1]}"
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: ⚠️ SyntaxError: inner double-quotes need to be replaced by single quotes.
image

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+                        "labels": [data.get("threat_type", None)],
+                    },
+                )
+                # create STIX indicator
+                pattern_value = "[ipv6-addr:value = '" + ip_value + "']"
+                observable_type = "IPv6-Addr"
+
+        else:
+            self.helper.log_warning(
+                f"Unrecognized ioc_type: {data.get("ioc_type", None)}"
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: ⚠️ SyntaxError: inner double-quotes need to be replaced by single quotes.
image

File: external-import/zvelo/src/zvelo_connector/utils.py [hunk 14]
```diff
@@ -0,0 +1,2 @@
+#  Utilities: helper functions, classes, or modules that provide common, reusable functionality across a codebase
+SUPPORTED_COLLECTIONS = ["phish", "threat", "malicious"]
```
Comment (Powlinett) [inline external-import/zvelo/src/zvelo_connector/utils.py:2]: If you need `SUPPORTED_COLLECTIONS` only in `connector.py` for config validation, I think you can move it to `config_variables.py` alongside `self.zvelo_collections` and remove `utils.py`.

File: external-import/zvelo/tests/test_zvelo_connector.py
Comment (Powlinett) [inline external-import/zvelo/tests/test_zvelo_connector.py:1]: Ideally, we'd like to have some basic unit tests here 😇 but if you don't want to add tests, then please remove `tests` directory to avoid committing "empty" files.

**Review by Powlinett — COMMENTED** [review]
Hi @romain-filigran, please find my feedback after some tests:

⚠️ Two SyntaxError lead the connector to crash instantly

Once these errors are fixed:
✅ connector's workflow is clear
✅ README is crystal clear too
✅ no errors during ingestion
⁉️after some runs I receive responses with status code `400 Bad Request`  for 'phish' collection only (see logs below - idk if it's intended)

Note that _**not** all my comments need to be addressed_, only the ones with warnings (⚠️). The other ones are just to share knowledge and/or quick win improvements.

Last but not least, I didn't write a specific comment about docstrings but can you fill/write them all please? 😇 I think this connector could benefit from later refactos and this type of comments could really help us to navigate through the code. It would be even better if you could also add some basic typing 🤩

Many thanks!

_________________________


400 Bad Request logs:
```
connector-zvelo-1  | {"timestamp": "2024-12-17T19:01:36.327422Z", "level": "ERROR", "name": "Zvelo", "message": "[API] Error while fetching data in collection 'phish': ", "exc_info": "Traceback (most recent call last):\n  File \"/opt/opencti-connector-zvelo/zvelo_connector/client_api.py\", line 78, in get_collections_entities\n    response.raise_for_status()\n  File \"/usr/local/lib/python3.12/site-packages/requests/models.py\", line 1024, in raise_for_status\n    raise HTTPError(http_error_msg, response=self)\nrequests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://api.zvelo.io/v1/phish?page=0&created_date_start=2024-12-17T18%3A00%3A41.047713%2B00%3A00\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/opt/opencti-connector-zvelo/zvelo_connector/connector.py\", line 162, in process_message\n    stix_objects = self._collect_intelligence(from_date=last_run)\n                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/opt/opencti-connector-zvelo/zvelo_connector/connector.py\", line 92, in _collect_intelligence\n    entities = self.client.get_collections_entities(\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/opt/opencti-connector-zvelo/zvelo_connector/client_api.py\", line 93, in get_collections_entities\n    raise Exception(error_msg)\nException: [API] Error while fetching data in collection 'phish': ", "taskName": null}
```

File: external-import/zvelo/src/zvelo_connector/connector.py (hunk 2 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/connector.py]: Done

File: external-import/zvelo/src/zvelo_connector/connector.py (hunk 3 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/connector.py]: Done

File: external-import/zvelo/src/zvelo_connector/connector.py (hunk 4 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/connector.py]: Done

File: external-import/zvelo/src/zvelo_connector/connector.py (hunk 5 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/connector.py]: Done

File: external-import/zvelo/src/zvelo_connector/connector.py (hunk 6 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/connector.py:143]: This is the format expected by the Zvelo API.

File: external-import/zvelo/src/zvelo_connector/connector.py (hunk 7 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/connector.py]: Done

File: external-import/zvelo/src/zvelo_connector/config_variables.py (hunk 8 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/config_variables.py]: Done

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py (hunk 9 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: Done

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py (hunk 10 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: Done

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py (hunk 11 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: Done

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py (hunk 12 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: Done

File: external-import/zvelo/src/zvelo_connector/converter_to_stix.py (hunk 13 above)
Comment (romain-filigran) [reply external-import/zvelo/src/zvelo_connector/converter_to_stix.py]: Done

File: external-import/zvelo/tests/test_zvelo_connector.py
Comment (romain-filigran) [reply external-import/zvelo/tests/test_zvelo_connector.py:1]: Done

**Comment by Powlinett** [comment]
✅ `400 - Bad request` was due to bad date format on my local branch
✅ I found a STIX pattern ValidationError due to unescaped single quotes in IOC urls so I pushed a fix

The PR is ready for merge 🙌

**Review by Powlinett — APPROVED** [review]
